File: lib/ashy_walnut_desk/ai/prompt.ex
defmodule AshyWalnutDesk.AI.Prompt do
  @moduledoc false

  @type system_block :: %{
          required(:type) => String.t(),
          required(:text) => String.t(),
          optional(:cache_control) => %{required(:type) => String.t()} | nil
        }

  @type message :: %{
          required(:role) => String.t(),
          required(:content) => String.t()
        }

  @type metadata :: map()

  @type t :: %__MODULE__{
          model: String.t() | nil,
          max_tokens: pos_integer() | nil,
          system_blocks: [system_block()],
          messages: [message()],
          metadata: metadata() | nil
        }

  defstruct [:model, :max_tokens, system_blocks: [], messages: [], metadata: %{}]
end

File: crates/awd-ai/src/prompt.rs
//! Prompt value types — the assembled request the Anthropic adapter serializes.
//! Mirrors `ai/prompt.ex`.

use chrono::{DateTime, Utc};

/// Inbound vs outbound message in the conversation transcript.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Direction {
    Inbound,
    Outbound,
}

/// A conversation message fed to the assembler. (A nil/unknown direction in the
/// Elixir source is treated as `Inbound`; construct it that way here.)
#[derive(Debug, Clone)]
pub struct Message {
    pub direction: Direction,
    pub body: String,
    pub inserted_at: Option<DateTime<Utc>>,
}

/// A system prompt block. `cache_control_ephemeral` forwards the
/// `cache_control: {type: "ephemeral"}` marker (set on framework + persona).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SystemBlock {
    pub text: String,
    pub cache_control_ephemeral: bool,
}

/// A chat message in the `messages` array.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

/// The assembled prompt.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Prompt {
    pub model: Option<String>,
    pub max_tokens: Option<u32>,
    pub system_blocks: Vec<SystemBlock>,
    pub messages: Vec<ChatMessage>,
    pub metadata: serde_json::Value,
}

/// Persona content used to build the persona system block.
#[derive(Debug, Clone, Default)]
pub struct Persona {
    pub system_prompt: Option<String>,
    pub guardrail_notes: Option<String>,
}
